text/0192-remove-constructs-compat.md

```diff
@@ -0,0 +1,888 @@
+
+This works in production but introduces a minor issue with unit tests which use
+a `Stack` as the root construct. This is a very common pattern in the CDK
+codebase today:
+
+```ts
+const stack = new Stack();
+const foo = new MyFoo(stack, 'MyFoo');
+
+expect(stack).to(haveResource('AWS::Foo'));
+```
+
+In such cases, assets will not work because the output directory is only
+determined later (in the `expect` call).
+
+One alternative would be to simply modify these unit tests so that stacks are no
+longer used as roots (basically add `new App()` as the scope). This approach
+would require a change in many unit tests across the code base with no clear
+value to users.
+
+We propose to modify the `Stack` construct such that if a stack is created
+without an explicit `scope`, an `App` instance will automatically be created and
+used as it's scope.
+
+```ts
+const stack = new Stack();
+assert(stack.node.scope instanceof App); // previously it was `undefined`
+```
+
+Since only the root construct may have an empty ID, we will also need to assign
+an ID. We propose to use `"Stack"` since we already have fallback logic that
+uses this as the stack name when the stack does not have an ID (see
+[stack.ts](https://github.com/aws/aws-cdk/blob/8c0142030dce359591aa76fe314f19fce9eddbe6/packages/%40aws-cdk/core/lib/stack.ts#L920)).
+
+This change will allow us to remove any special casing we have for stacks in the
+testing framework and throughout the synthesis code path (we have quite a lot of
+that), because we will be able to assume that `Stage.of(construct.node.root)` is
+never `undefined` and has a `synth()` method which returns a cloud assembly.
+
+Unit tests sometimes use "incremental tests" for synthesized templates. For
+example:
+
+```ts
+const stack = new Stack();
+const c1 = new MyConstruct(stack, 'c1', { foos: [ 'bar' ] });
+expect(stack).toHaveResource('AWS::Resource', {
+  Foos: [ 'bar' ]
+});
+
+// now add a "foo" and verify that the synthesized template contains two items
+c1.addFoo('baz');
+expect(stack).toHaveResource('AWS::Resource', {
+  Foos: [ 'bar', 'baz' ]
+});
+```
+
+Since `stage.synth()` (which is called by `expect(stack)`) would reuse the
+synthesized output if called twice, we will also need to introduce a
+`stage.synth({ force: true })` option. This will be the default behavior when
+using `expect(stack)` or `SynthUtils.synth()`.
+
+### Preserving paths and unique IDs using `node.relocate()`
+
+A side effect of adding a `App` parent to "root" stacks is that construct paths
+and unique IDs in unit tests will now change. In the above example,
+`foo.node.path` will change from `MyFoo` to `Stack/MyFoo`. Additionally, tests
+for resources that utilized `node.uniqueId` to generate names will also change
+given `uniqueId` is based on the path.
+
+
+To address this, we propose to introduce a new capability in `constructs` to
+"**relocate**" a scope to a different path. This capability will also be useful
```

Comment:
Sounds dangerous as a feature available to customers.
Maybe more dangerous than stack mutation during `prepare()`?

Reply:
Reverted and decided to use `Default` as the name of the root stack, which is already ignored by `uniqueId` yey!